### PR TITLE
feat: upload image to aws s3

### LINKED ISSUE #152: Upload photos to Amazon S3
## Description
As a user of Mentorship System,
I need to be able to upload profile image into the backend,
so that I can make my profile more personal.

You can find some info about S3 urls in [this file of Android app](https://github.com/systers/mentorship-android/blob/develop/app/src/main/java/org/systers/mentorship/remote/BaseUrl.kt).

**Expected Input:** Image will be sent as an encoded string

### Update [Required]
- [ ] Save User images on Amazon S3
- [ ] Save URL on UserModel `photo_url` field
- [ ] Return `photo_url` on appropriate APIs

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
5 hours

### PR DESCRIPTION
**Pull request by CleverFool77**
### Description
- [x] Save User images on Amazon S3
- [x] Save object URL of image in aws s3 on UserModel photo_url field
- [x] Return photo_url on new api endpoint user/profile_picture

Fixes #152

Server will be getting an encoded string of image as input,The image will be decoded and uploaded on aws s3. Then aws s3 url of the image will be saved in the db field of user's photo_url. A new api endpoint to return the user's profile picture is created.
The testing has been done in my own aws s3 bucket. Three environment variables `AWS_ACCESS_KEY_ID` , `AWS_SECRET_ACCESS_KEY` and `AWS_PROFILE_BUCKET` are created. The username is taken as image name to avoid multiple images of user.
I've extracted the image extension from the encoded string input (uploaded from frontend).

It's assumed that the frontend encodes the image in base64 encoding.

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
This has been tested using postman.

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged
- [x] Update Postman API at /docs folder
- [x] Update Swagger documentation and the exported file at /docs folder
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by CleverFool77** [comment]
@isabelcosta @ramitsawhney27 I've opened a Pr reagrding the upload of image in amazon s3 for the issue #152

File: app/api/dao/user.py
Comment (agarwalrounak) [inline app/api/dao/user.py:13]: Add one more blank line before class definition. Refer: https://www.python.org/dev/peps/pep-0008/#blank-lines

Comment (CleverFool77) [reply app/api/dao/user.py:13]: @ramitsawhney27 @agarwalrounak I'm little confused about which linting tool we are using. It's pep8 , flake8 or pylint ?

Comment (ramitsawhney27) [reply app/api/dao/user.py:13]: pylint

File: app/api/dao/user.py
Comment (ramitsawhney27) [inline app/api/dao/user.py]: username + "." + image_type
Need to add the spaces between operators, and maybe a better way to handle this string append.

File: app/api/dao/user.py
Comment (ramitsawhney27) [inline app/api/dao/user.py]: not pythonic, a better way is:
try:
image_type = ...
except KeyError:
<handle>

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]

File: app/api/dao/user.py
Comment (ramitsawhney27) [inline app/api/dao/user.py]: incorrect import order

Comment (CleverFool77) [reply app/api/dao/user.py]: @ramitsawhney27 Does it mean shift `import os` to top ?

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: @ramitsawhney27  This practice is followed in whole function, Should I open an issue and fix it in another PR ?

File: app/api/dao/user.py
Comment (agarwalrounak) [reply app/api/dao/user.py]: I think yes. Refer https://www.python.org/dev/peps/pep-0008/#imports and https://stackoverflow.com/questions/20762662/whats-the-correct-way-to-sort-python-import-x-and-from-x-import-y-statement

Comment (CleverFool77) [reply app/api/dao/user.py]: @agarwalrounak Thanks for the link :)

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py:13]: @ramitsawhney27 Thanks for the info :)

File: app/api/dao/user.py
Comment (ramitsawhney27) [reply app/api/dao/user.py]: Lets just fix it here for now

Comment (CleverFool77) [reply app/api/dao/user.py]: Although I haven't touched that line. If I'll do the changes it won't be a unnecessary change for the PR right. @ramitsawhney27

Comment (ramitsawhney27) [reply app/api/dao/user.py]: That's not an issue.

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: Here, i'm thinking of assigning a new variable with username + "." + image_type and using variable as a parameter in function.

**Comment by CleverFool77** [comment]
@ramitsawhney27 @agarwalrounak I've updated the PR.

File: app/api/dao/user.py
Comment (ramitsawhney27) [inline app/api/dao/user.py]: join is better here.

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: image_name = ".".join((username, image_type)) How about this.

**Comment by CleverFool77** [comment]
@ramitsawhney27 I've updated the PR with the changes and removed the merge conflict.

**Comment by isabelcosta** [comment]
@CleverFool77 uploading an image is testable on Postman? I have to try this. Thank you so much for the work you're doing here.

**Comment by CleverFool77** [comment]
@isabelcosta Yes we can test it on Postman.
* In the code, we have to setup the environment variable which include the name of bucket and other aws details.
* As the input image in encoded string , so in update user profile endpoint. we put the encoded string in base64 of image as input.
* The further process will be according to code - Uploading decoded image to aws s3, storing the image url from aws s3 to database.
* The link of user's profile image url in amazon aws s3 is returned in Profile picture endpoint.

**Review by ramitsawhney27 — DISMISSED** [review]

File: app/api/dao/user.py
Comment (isabelcosta) [inline app/api/dao/user.py]: Can you import this environment variable on [develop/config.py](https://github.com/systers/mentorship-backend/blob/develop/config.py) file, so that we use it here like the rest of the environments... `àpplication.config.ENV_VAR`

File: app/api/dao/user.py
Comment (isabelcosta) [inline app/api/dao/user.py]: Now that you're other PR was merged can you make this message a constant ?

**Review by isabelcosta — COMMENTED** [review]
@CleverFool77 I left two comments. I haven't got the time to test this feature yet. Once I do, I'll give you a complete review

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: @isabelcosta Sure :)

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: I'll update this.

**Comment by isabelcosta** [comment]
@m-murad we have to setup these (e.g.: AWS_ACCESS_KEY and etc) on TravisCI right?
We need @MayBurgos help with the credentials as well.

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@CleverFool77 I'm ok with this after you make the changes I requested ;)

**Comment by CleverFool77** [comment]
Hi @isabelcosta I've updated the PR as requested. 
Thanks :)

**Review by ramitsawhney27 — DISMISSED** [review]

File: app/api/dao/user.py
Comment (isabelcosta) [inline app/api/dao/user.py]:
```suggestion
                photo_url = aws_upload.upload_to_aws(data['photo_url'], application.config["AWS_PROFILE_BUCKET"], image_name)
```

File: config.py
Comment (isabelcosta) [inline config.py:20]: Bravo ;) this is exactly what I meant 🎉

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@CleverFool77 left a comment for a typo. It's almost done ;)

**Comment by CleverFool77** [comment]
Hi @isabelcosta
I missed out the mention.
I've updated the PR.
Thanks for the patience !!

File: app/api/dao/user.py
Comment (CleverFool77) [reply app/api/dao/user.py]: Updated .

**Review by ramitsawhney27 — DISMISSED** [review]

**Review by isabelcosta — APPROVED** [review]

**Comment by isabelcosta** [comment]
@CleverFool77 thank you so much for your contribution and patience! Well done! 🎉 

I'm not sure when I will merge this because I haven't figure out yet if it's better to set up the environments to the AWS Systers uses, or not. I'm also analyzing the volatile database issue and where data will be stored, and all of this thinking about the development environment. Your work here is done, just wanted to let you know where I'm at.

There's also the possibility of merging this without the environment variables set, and all run ok. I have to test this, yet. Once I do, if all goes well I can merge right away.

 cc @ramitsawhney27

File: config.py
Comment (isabelcosta) [inline config.py]:
```suggestion
    AWS_ACCESS_KEY_ID = os.getenv('AWS_ACCESS_ID')
```

Can you please change the expected environment variable to `AWS_ACCESS_ID`. I'm saying this because during the Community Open Session with @MayBurgos and @m-murad that Travis CI has that variable like `AWS_ACCESS_ID` instead of `AWS_ACCESS_KEY_ID`.

File: app/api/dao/user.py
Comment (isabelcosta) [reply app/api/dao/user.py]: @CleverFool77 this doesn't appear to be changed. Maybe it happened something when you squashed the commits. Can you please change that.
```suggestion
                photo_url = aws_upload.upload_to_aws(data['photo_url'], application.config["AWS_PROFILE_BUCKET"], image_name)
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@CleverFool77 left some comments because of something I noticed on the Travis CI current configuration.
Let me know if you need any clarification.

**Comment by CleverFool77** [comment]
Thanks @isabelcosta for the review. 
I'll update the PR soon.

**Comment by CleverFool77** [comment]
Hii @isabelcosta !!
I have updated the PR as requested.

Thanks !!!

**Comment by vj-codes** [comment]
@CleverFool77 hey can you please solve the merge conflicts ? Thanks!

**Comment by vj-codes** [comment]
@CleverFool77 closing the PR due to inactivity , thank you for your contribution :)

**Comment by isabelcosta** [comment]
@vj-codes thank you for closing this. At the moment we don't have access to AWS S3 so this wouldn't be merged so soon. We'll reopen if the situation changes.
Thank you for your contribution @CleverFool77 , I hope you learned something new with this PR, I sure did :)

**Pull request closed without merging**
